threshsign/src/eddsa/EdDSAMultisigVerifier.cpp

```diff
@@ -0,0 +1,130 @@
+  ConcordAssertGE(static_cast<uint64_t>(threshSigLen),
+                  static_cast<unsigned long>(signatures_.size()) * sizeof(SingleEdDSASignature));
+  size_t offset = 0;
+  for (auto &[id, sig] : signatures_) {
+    UNUSED(id);
+    std::memcpy(outThreshSig + offset, &sig, sizeof(SingleEdDSASignature));
+    offset += sizeof(SingleEdDSASignature);
+  }
+  return offset;
+}
+
+bool EdDSASignatureAccumulator::hasShareVerificationEnabled() const { return verification_; }
+int EdDSASignatureAccumulator::getNumValidShares() const {
+  return static_cast<int>(signatures_.size() - invalidShares_.size());
+}
+std::set<ShareID> EdDSASignatureAccumulator::getInvalidShareIds() const { return invalidShares_; }
+
+EdDSASignatureAccumulator::EdDSASignatureAccumulator(bool verification, const EdDSAMultisigVerifier &verifier)
+    : verification_(verification), verifier_(verifier) {
+  LOG_DEBUG(EDDSA_MULTISIG_LOG, KVLOG(this));
+}
+
+EdDSAMultisigVerifier::EdDSAMultisigVerifier(const std::vector<SingleVerifier> &verifiers,
+                                             const size_t signersCount,
+                                             const size_t threshold)
+    : verifiers_(verifiers), signersCount_(signersCount), threshold_(threshold) {
+  ConcordAssertEQ(verifiers.size(), signersCount + 1);
+  LOG_DEBUG(EDDSA_MULTISIG_LOG, KVLOG(this, verifiers_.size(), threshold_));
+}
+
+IThresholdAccumulator *EdDSAMultisigVerifier::newAccumulator(bool withShareVerification) const {
+  return new EdDSASignatureAccumulator(withShareVerification, *this);
+}
+
+bool EdDSAMultisigVerifier::verifySingleSignature(const concord::Byte *msg,
+                                                  size_t msgLen,
+                                                  const SingleEdDSASignature &signature) const {
+  if (signature.id == 0) {
+    return false;
+  }
+  return verifiers_[signature.id].verifyBuffer(
+      msg, msgLen, signature.signatureBytes.data(), signature.signatureBytes.size());
+}
+
+bool EdDSAMultisigVerifier::verify(const char *msg, int msgLen, const char *sig, int sigLen) const {
+  LOG_DEBUG(EDDSA_MULTISIG_LOG, KVLOG(this, signersCount_, threshold_, sigLen));
+  auto msgLenUnsigned = static_cast<size_t>(msgLen);
+  auto sigLenUnsigned = static_cast<unsigned long>(sigLen);
+  ConcordAssert(sigLenUnsigned % sizeof(SingleEdDSASignature) == 0);
+  const auto signatureCountInBuffer = sigLenUnsigned / sizeof(SingleEdDSASignature);
+
+  if (signatureCountInBuffer < threshold_) {
+    return false;
+  }
+
+  const SingleEdDSASignature *allSignatures = reinterpret_cast<const SingleEdDSASignature *>(sig);
+  size_t validSignatureCount = 0;
+
+  for (int i = 0; i < static_cast<int>(signatureCountInBuffer); i++) {
+    auto &currentSignature = allSignatures[i];
+    if (currentSignature.id == 0 || currentSignature.id >= verifiers_.size()) {
+      LOG_ERROR(EDDSA_MULTISIG_LOG, "Invalid signer id" << KVLOG(currentSignature.id, verifiers_.size()));
+      continue;
```

Comment:
Is this a valid case to just continue? I am not "in the loop" on where is the signer id coming from, if it comes from network/storage then we shouldnt terminate - else everyone who is byzantine autehticated can crash the replica. if it's some internal error maybe termination is better.